fast_llm/tools/prepare_dataset.py

```diff
@@ -0,0 +1,359 @@
+import abc
+import argparse
+import json
+import os
+import pathlib
+import shutil
+import typing
+from multiprocessing import Pool
+
+import numpy as np
+import torch.distributed
+
+from fast_llm.config import Config, Field, FieldHint, check_field, config_class
+from fast_llm.data.config import TokenizerConfig
+from fast_llm.data.gpt.memmap import GPTMemmapDataset
+from fast_llm.data.tokenizer import Tokenizer
+from fast_llm.engine.config_utils.data_type import DataType
+from fast_llm.engine.config_utils.runnable import RunnableConfig
+from fast_llm.utils import Assert, Registry
+
+
+@config_class
+class DistributedConfig(Config):
+    default_world_size: typing.ClassVar[int] = int(os.environ.get("WORLD_SIZE", 1))
+    default_rank: typing.ClassVar[int] = int(os.environ.get("RANK", 0))
+    world_size: int = Field(
+        default=None,
+        desc="Size of the world group. Typically provided by torchrun or equivalent through the `WORLD_SIZE` environment variable.",
+        hint=FieldHint.expert,
+        valid=check_field(Assert.gt, 0),
+    )
+    rank: int = Field(
+        default=None,
+        desc="Rank of the local process. Typically provided by torchrun or equivalent through the `RANK` environment variable.",
+        hint=FieldHint.expert,
+        valid=check_field(Assert.geq, 0),
+    )
+    backend: str = Field(
+        default="gloo",
+        desc="Distributed backend to use.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.incl, torch.distributed.Backend.backend_list),
+    )
+
+    def _validate(self):
+        if self.world_size is None:
+            self.world_size = self.default_world_size
+        if self.rank is None:
+            self.rank = self.default_rank
+        super()._validate()
+        Assert.in_range(self.rank, 0, self.world_size)
+
+
+@config_class()
+class DatasetPreparatorConfig(RunnableConfig):
```

Comment:
Could we move dataset preparators to a separate module? Maybe `fast_llm/data/prepare/`? (and put config and gpt in separate files)

Reply:
Great, yes! I was hoping you'd tell me where to put this